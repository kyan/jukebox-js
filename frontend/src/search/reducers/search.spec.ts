import reducer, { SearchState } from './search'
import Constant from 'search/constants'

const getInitialState = (initial?: Partial<SearchState>): any =>
  reducer(initial as SearchState, {} as any)

describe('search', () => {
  it('handles initial state', () => {
    const initialState = getInitialState()
    expect(initialState).toMatchSnapshot()
  })

  it('handles REMOVE_FROM_SEARCH_RESULTS', () => {
    const initialState = getInitialState({
      results: [{ track: { uri: 'url1' } }, { track: { uri: 'url2' } }, { track: { uri: 'url3' } }]
    })
    expect(initialState.results).toHaveLength(3)
    const state = reducer(initialState, {
      type: Constant.REMOVE_FROM_SEARCH_RESULTS,
      uris: ['url2']
    })
    expect(state.results).toHaveLength(2)
    expect(state.results[0].track.uri).toEqual('url1')
  })

  it('handles TOGGLE_SEARCH_SIDEBAR', () => {
    const initialState = getInitialState()
    const state = reducer(initialState, {
      type: Constant.TOGGLE_SEARCH_SIDEBAR,
      open: true
    })
    expect(state.searchSideBarOpen).toBeTruthy()
  })

  it('handles STORE_SEARCH_QUERY', () => {
    const initialState = getInitialState()
    const state = reducer(initialState, {
      type: Constant.STORE_SEARCH_QUERY,
      query: 'query'
    })
    expect(state.query).toEqual('query')
  })

  it('handles STORE_SEARCH_RESULTS', () => {
    const tracks = [
      { track: { uri: 'url1' } },
      { track: { uri: 'url2' } },
      { track: { uri: 'url3' } }
    ]
    const initialState = getInitialState()
    expect(initialState.results).toHaveLength(0)
    const state = reducer(initialState, {
      type: Constant.STORE_SEARCH_RESULTS,
      results: {
        tracks: {
          offset: 0,
          total: 3,
          items: tracks
        }
      }
    })
    expect(state.results).toHaveLength(3)
    expect(state.offset).toEqual(0)
    expect(state.total).toEqual(3)
  })

  it('handles STORE_SEARCH_RESULTS of > 30000', () => {
    const tracks = [
      { track: { uri: 'url1' } },
      { track: { uri: 'url2' } },
      { track: { uri: 'url3' } }
    ]
    const initialState = getInitialState()
    expect(initialState.results).toHaveLength(0)
    const state = reducer(initialState, {
      type: Constant.STORE_SEARCH_RESULTS,
      results: {
        tracks: {
          offset: 0,
          total: 30000,
          items: tracks
        }
      }
    })
    expect(state.results).toHaveLength(3)
    expect(state.offset).toEqual(0)
    expect(state.total).toEqual(10000)
  })
})
