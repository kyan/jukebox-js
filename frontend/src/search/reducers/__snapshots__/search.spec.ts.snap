// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search handles initial state 1`] = `
Object {
  "limit": 20,
  "offset": 0,
  "query": "",
  "results": Array [],
  "searchSideBarOpen": false,
  "total": 0,
}
`;
